src/audio/sources/mp3.cpp

```diff
@@ -0,0 +1,194 @@
+ */
+
+#include <cassert>
+#include <cstdint>
+#include <cstdio>
+#include <cstdlib>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <sstream>
+#include <string>
+
+extern "C" {
+#include <mpg123.h>
+}
+
+#include "../../errors.hpp"
+#include "../../messages.h"
+#include "../audio_source.hpp"
+#include "../sample_formats.hpp"
+#include "mp3.hpp"
+
+// This value is somewhat arbitrary, but corresponds to the minimum buffer size
+// used by ffmpeg, so it's probably sensible.
+const size_t Mp3AudioSource::BUFFER_SIZE = 16384;
+
+/* static */ std::unique_ptr<AudioSource> Mp3AudioSource::Build(
+                const std::string &path)
+{
+	return std::unique_ptr<AudioSource>(new Mp3AudioSource(path));
+}
+
+Mp3AudioSource::Mp3AudioSource(const std::string &path)
+    : AudioSource(path), buffer(BUFFER_SIZE), context(nullptr)
+{
+	this->context = mpg123_new(nullptr, nullptr);
+	mpg123_format_none(this->context);
+
+	const long *rates = nullptr;
+	size_t nrates = 0;
+	mpg123_rates(&rates, &nrates);
+	for (size_t r = 0; r < nrates; r++) {
+		Debug() << "trying to enable formats at " << rates[r]
+		        << std::endl;
+		AddFormat(rates[r]);
+	}
+
+	if (mpg123_open(this->context, path.c_str()) == MPG123_ERR) {
+		throw FileError("mp3: can't open " + path + ": " +
+		                mpg123_strerror(this->context));
+	}
+
+	Debug() << "mp3: sample rate:" << this->SampleRate() << std::endl;
+	Debug() << "mp3: bytes per sample:" << this->BytesPerSample()
+	        << std::endl;
+	Debug() << "mp3: channels:" << (int)this->ChannelCount() << std::endl;
+	Debug() << "mp3: playd format:" << (int)this->OutputSampleFormat()
+	        << std::endl;
+}
+
+Mp3AudioSource::~Mp3AudioSource()
+{
+	mpg123_delete(this->context);
+	this->context = nullptr;
+}
+
+void Mp3AudioSource::AddFormat(long rate)
+{
+	// The requested encodings correspond to the sample formats available in
+	// the SampleFormat enum.
+	if (mpg123_format(this->context, rate, MPG123_STEREO | MPG123_MONO,
+	                  (MPG123_ENC_UNSIGNED_8 | MPG123_ENC_SIGNED_8 |
+	                   MPG123_ENC_SIGNED_16 | MPG123_ENC_SIGNED_32 |
+	                   MPG123_ENC_FLOAT_32)) == MPG123_ERR) {
+		// Ignore the error for now -- another sample rate may be
+		// available.
+		// If no sample rates work, loading a file will fail anyway.
+		Debug() << "can't support" << rate << std::endl;
+	};
+}
+
+std::uint8_t Mp3AudioSource::ChannelCount() const
+{
+	assert(this->context != nullptr);
+
+	int chans = 0;
+	mpg123_getformat(this->context, nullptr, &chans, nullptr);
+	assert(chans != 0);
+	return static_cast<std::uint8_t>(chans);
+}
+
+std::uint32_t Mp3AudioSource::SampleRate() const
+{
+	assert(this->context != nullptr);
+
+	long rate = 0;
+	mpg123_getformat(this->context, &rate, nullptr, nullptr);
+
+	assert(0 < rate);
+	// INT32_MAX isn't a typo; if we compare against UINT32_MAX, we'll
+	// set off sign-compare errors, and the sample rate shouldn't be above
+	// INT32_MAX anyroad.
```

Comment:
huh?  long is signed and possibly 64 bits long, so the asserts are there to ensure that rate is in the range 0..INT32_MAX and the uint32_t cast is sound.  The comment is saying that we would normally check 0..UINT32_MAX, but this throws sign comparison warnings/errors.